lsp-completion.el

```diff
@@ -0,0 +1,548 @@
+  (unwind-protect
+      (-let* (((&plist 'lsp-completion-item item
+                       'lsp-completion-start-point start-point
+                       'lsp-completion-markers markers
+                       'lsp-completion-prefix prefix)
+               (text-properties-at 0 candidate))
+              ((&CompletionItem :label :insert-text? :text-edit? :insert-text-format? :additional-text-edits?)
+               item))
+        (cond
+         (text-edit?
+          (apply #'delete-region markers)
+          (insert prefix)
+          (lsp--apply-text-edit text-edit?))
+         ((or insert-text? label)
+          (apply #'delete-region markers)
+          (insert prefix)
+          (delete-region start-point (point))
+          (insert (or insert-text? label))))
+
+        (when (eq insert-text-format? 2)
+          (let ((yas-indent-line (lsp--indent-snippets?)))
+            (yas-expand-snippet
+             (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
+             start-point
+             (point))))
+
+        (when lsp-completion-enable-additional-text-edit
+          (if (or (get-text-property 0 'lsp-completion-resolved candidate)
+                  additional-text-edits?)
+              (lsp--apply-text-edits additional-text-edits?)
+            (-let [(callback cleanup-fn) (lsp--create-apply-text-edits-handlers)]
+              (lsp--resolve-completion-async
+               item
+               (lambda (resolved-item)
+                 (funcall callback
+                          (lsp:completion-item-additional-text-edits? resolved-item)))
+               cleanup-fn))))
+
+        (when (and lsp-signature-auto-activate
+                   (lsp-feature? "textDocument/signatureHelp"))
+          (lsp-signature-activate))
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+;;;###autoload
+(defun lsp-completion--enable ()
+  "Enable LSP completion support."
+    (lsp-completion-mode 1))
+
+(defun lsp-completion--disable ()
+  "Disable LSP completion support."
+  (lsp-completion-mode -1))
+
+;;;###autoload
+(define-minor-mode lsp-completion-mode
+  "Toggle LSP completion support."
+  :group 'lsp-mode
+  :global nil
+  :lighter ""
+  (cond
+   (lsp-completion-mode
+    (when (and lsp-completion-enable
+               (lsp-feature? "textDocument/completion"))
+      (setq-local completion-at-point-functions nil)
+      (add-hook 'completion-at-point-functions #'lsp-completion-at-point nil t)
+      (setq-local completion-category-defaults
+                  (add-to-list 'completion-category-defaults '(lsp-capf (styles basic)))))
+
+    (when (lsp--capability :completionProvider)
+      (cond
+       ((and (functionp 'company-lsp)
+             (not lsp-completion-prefer-capf))
+        (progn
+          (company-mode 1)
```

Comment:
I would say this is auto configure code, since it messes directly with user configuration.
For example, I have `company-mode`  installed and not planning to use it, instead using completion via `ivy-posframe`, this will override that configuration.

Reply:
sorry, I don't get it, what's your suggestion?

Reply:
I mean the part of 
``` elisp
(cond
       ((and (functionp 'company-lsp)
             (not lsp-completion-prefer-capf))
        (progn
          (company-mode 1)
...
```
That's user configuration since it modifies `company-backends`. I have a special configuration for `capf` (like multi backend for `company-capf` and `dabbrev-code`) and by this, I will have no way of keeping my `company-backends` unchanged. It will also break my configuration.
I think this should move to `lsp-completion--auto-configuration` function like before so user can disable auto configuration for `lsp-completion` (which is set up `company-mode` and enable `lsp-completion-mode`?) and use their own initialization code for it.

Reply:
It makes sense, done!
I just don't understand how you should be able to have other features auto enabled as if you disable `lsp-auto-configure` because of your custom company-backends it will disable LSP auto configuration for all LSP features as well.

Reply:
I disable auto configure and only enable it for some features.